crates/red_knot_python_semantic/src/types.rs

```diff
@@ -762,12 +762,25 @@ impl<'db> CallOutcome<'db> {
             } => {
                 let mut not_callable = vec![];
                 let mut union_builder = UnionBuilder::new(db);
+                let mut revealed = false;
                 for outcome in &**outcomes {
-                    let return_ty = if let Self::NotCallable { not_callable_ty } = outcome {
-                        not_callable.push(*not_callable_ty);
-                        Type::Unknown
-                    } else {
-                        outcome.unwrap_with_diagnostic(db, node, builder)
+                    let return_ty = match outcome {
+                        Self::NotCallable { not_callable_ty } => {
+                            not_callable.push(*not_callable_ty);
+                            Type::Unknown
+                        }
+                        Self::RevealType {
+                            return_ty,
+                            revealed_ty: _,
+                        } => {
+                            if revealed {
+                                *return_ty
+                            } else {
+                                revealed = true;
+                                outcome.unwrap_with_diagnostic(db, node, builder)
+                            }
+                        }
+                        _ => outcome.unwrap_with_diagnostic(db, node, builder),
```

Comment:
I wonder if this demonstrates a more general problem where we probably don't want to emit one diagnostic for each union member if some of them are not callable. 

For example type script groups the error:

```ts
let a: number | string;

a();
```

```
This expression is not callable.
  No constituent of type 'string | number' is callable.(2349)
```

and I think that's better than having multiple diagnostics with the same range.

Reply:
I understand the idea, but I'm confused by the comment, because I think the code here already implements exactly what you're asking for? We already consolidate all not-callable errors from calling a union and emit just a single diagnostic naming all the not-callable elements of the union.

This wasn't implemented in this PR, it was implemented in the original PR that added union-call support: https://github.com/astral-sh/ruff/pull/13384